File: app/src/main/java/site/doramusic/app/ui/activity/BrowserActivity.java
package site.doramusic.app.ui.activity;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.KeyEvent;
import android.view.ViewGroup;
import android.webkit.JavascriptInterface;
import android.widget.LinearLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.just.agentweb.AgentWeb;
import com.just.agentweb.WebIndicator;

import dora.BaseActivity;
import dora.util.IntentUtils;
import dora.util.StatusBarUtils;
import site.doramusic.app.R;
import site.doramusic.app.databinding.ActivityBrowserBinding;

public class BrowserActivity extends BaseActivity<ActivityBrowserBinding> {

    private String title = "扫码结果";
    private String url;
    private AgentWeb agentWeb;

    @Override
    protected int getLayoutId() {
        return R.layout.activity_browser;
    }

    @Override
    protected void onGetExtras(@Nullable String action, @Nullable Bundle bundle, @NonNull Intent intent) {
        if (IntentUtils.hasExtra(intent, "title")) {
            title = IntentUtils.getStringExtra(intent, "title");
        }
        if (IntentUtils.hasExtra(intent, "url")) {
            url = IntentUtils.getStringExtra(intent, "url");
        }
    }

    @Override
    protected void onSetStatusBar() {
        StatusBarUtils.setStatusBarColorRes(this, R.color.colorPrimaryDark);
    }

    @Override
    public void initData(Bundle savedInstanceState) {
        mBinding.titlebar.setTitle(title);
        WebIndicator webIndicator = new WebIndicator(this);
        webIndicator.setColor(getResources().getColor(R.color.colorPrimary));
        agentWeb = AgentWeb.with(this)
                .setAgentWebParent(mBinding.rlBrowserWebPage,
                        new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,
                                ViewGroup.LayoutParams.MATCH_PARENT)
                )
                .setCustomIndicator(webIndicator)
                .createAgentWeb()
                .ready()
                .go(url);
        agentWeb.getJsInterfaceHolder().addJavaObject("android", new AndroidInterface(agentWeb, this, new WebJsInterfaceCallback() {
            @Override
            public void nativeApi(String params) {

            }
        }));
    }

    @Override
    public void onPointerCaptureChanged(boolean hasCapture) {
    }

    public interface WebJsInterfaceCallback {

        void nativeApi(String params);

    }

    public class AndroidInterface {
        private AgentWeb agent;
        private Context context;
        private WebJsInterfaceCallback interfaceCallback;
        private String TAG = "AndroidInterfaceWeb";

        public AndroidInterface(AgentWeb agent, Context context, WebJsInterfaceCallback interfaceCallback) {
            this.agent = agent;
            this.context = context;
            this.interfaceCallback = interfaceCallback;
        }

        //定义h5要调用的本地方法
        @JavascriptInterface
        public void Android_BuyVip(String mtdName, String params) {
            switch (mtdName) {
                case "nativeApi":
                    if (interfaceCallback != null) {
                        interfaceCallback.nativeApi(params);
                    }
                    break;
            }
        }
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        return agentWeb.handleKeyEvent(keyCode, event);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        agentWeb.destroy();
    }
}

File: app/src/main/java/site/doramusic/app/ui/activity/BrowserActivity.kt
package site.doramusic.app.ui.activity

import android.content.Intent
import android.content.res.Configuration
import android.os.Build
import android.os.Bundle
import android.view.KeyEvent
import android.view.View
import androidx.appcompat.widget.AppCompatImageView
import com.alibaba.android.arouter.facade.annotation.Route
import com.just.agentweb.AgentWeb
import com.just.agentweb.WebIndicator
import dora.skin.SkinManager
import dora.skin.base.BaseSkinBindingActivity
import dora.util.IntentUtils
import dora.util.StatusBarUtils
import dora.util.ViewUtils
import dora.widget.DoraTitleBar
import site.doramusic.app.R
import site.doramusic.app.base.conf.ARoutePath
import site.doramusic.app.base.conf.AppConfig.Companion.COLOR_THEME
import site.doramusic.app.base.conf.AppConfig.Companion.EXTRA_TITLE
import site.doramusic.app.base.conf.AppConfig.Companion.EXTRA_URL
import site.doramusic.app.databinding.ActivityBrowserBinding

/**
 * 浏览器页面，主要用于加载用户协议和隐私权政策等。
 */
@Route(path = ARoutePath.ACTIVITY_BROWSER)
class BrowserActivity : BaseSkinBindingActivity<ActivityBrowserBinding>() {

    private var title: String? = null
    private var url: String? = null
    private var agentWeb: AgentWeb? = null

    override fun getLayoutId(): Int {
        return R.layout.activity_browser
    }

    override fun onGetExtras(action: String?, bundle: Bundle?, intent: Intent) {
        title = IntentUtils.getStringExtra(intent, EXTRA_TITLE)
        url = IntentUtils.getStringExtra(intent, EXTRA_URL)
    }

    override fun onSetStatusBar() {
        val skinThemeColor = SkinManager.getLoader().getColor(COLOR_THEME)
        StatusBarUtils.setStatusBarColor(this, skinThemeColor)
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        ViewUtils.showSystemBar(window)
    }

    override fun initData(savedInstanceState: Bundle?, binding: ActivityBrowserBinding) {
        val skinThemeColor = SkinManager.getLoader().getColor(COLOR_THEME)
        binding.titlebar.setBackgroundColor(skinThemeColor)
        title?.let { binding.titlebar.title = it }
        val webIndicator = WebIndicator(this)
        webIndicator.setColor(skinThemeColor)
        binding.titlebar.addMenuButton(R.drawable.ic_min)
        binding.titlebar.setOnIconClickListener(object : DoraTitleBar.OnIconClickListener {
            override fun onIconBackClick(icon: AppCompatImageView) {
            }

            override fun onIconMenuClick(position: Int, icon: AppCompatImageView) {
                // 画中画模式
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    enterPictureInPictureMode()
                }
            }
        })
    }

    override fun onPictureInPictureModeChanged(isInPictureInPictureMode: Boolean, newConfig: Configuration) {
        super.onPictureInPictureModeChanged(isInPictureInPictureMode, newConfig)
        if (isInPictureInPictureMode) {
            mBinding.titlebar.visibility = View.GONE
        } else {
            mBinding.titlebar.visibility = View.VISIBLE
        }
    }

    override fun onPointerCaptureChanged(hasCapture: Boolean) {}

    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        return agentWeb?.handleKeyEvent(keyCode, event) ?: super.onKeyDown(keyCode, event)
    }

    override fun onDestroy() {
        super.onDestroy()
        agentWeb?.destroy()
    }
}